flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -983,8 +983,23 @@ struct EmboxCommonConversion : public FIROpConversion<OP> {
                     mlir::ConversionPatternRewriter &rewriter) const {
     auto thisPt = rewriter.saveInsertionPoint();
     auto *thisBlock = rewriter.getInsertionBlock();
-    auto func = mlir::cast<mlir::LLVM::LLVMFuncOp>(thisBlock->getParentOp());
-    rewriter.setInsertionPointToStart(&func.front());
+    auto op = thisBlock->getParentOp();
+    // Order to find the Op in whose entry block the alloca should be inserted.
+    // The parent Op if it is an LLVM Function Op.
+    // The ancestor OpenMP Op which is outlineable.
+    // The ancestor LLVM Function Op.
+    if (isa<mlir::LLVM::LLVMFuncOp>(op)) {
+      auto func = mlir::cast<mlir::LLVM::LLVMFuncOp>(op);
+      rewriter.setInsertionPointToStart(&func.front());
+    } else if (auto iface =
+                   thisBlock->getParent()
+                       ->getParentOfType<
+                           mlir::omp::OutlineableOpenMPOpInterface>()) {
+      rewriter.setInsertionPointToStart(iface.getAllocaBlock());
+    } else {
+      auto func = op->getParentOfType<mlir::LLVM::LLVMFuncOp>();
+      rewriter.setInsertionPointToStart(&func.front());
+    }
```

Comment:
I seems like the 1st and 3rd cases can be combined since they are ostensibly doing the same thing.

Also, should the second case be more general? Maybe OutlineableOpenMPOpInterface has some Op property that means, "this op captures variables"?